File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binarySearch(is_first):
            l , r = 0 , len(nums) - 1
            idx = -1

            while l <= r:
                m = (l+r) //2
                if nums[m] == target:
                    idx = m
                    if is_first :
                        r = m - 1
                    else:
                        l = m+1
                elif nums[m] < target:
                    l = m +1
                else:
                    r = m -1
            return idx
        return [binarySearch(True) , binarySearch(False)]

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
private:
int binarySearch(vector<int>&nums ,int target, bool is_true){
    int l = 0 , r = nums.size() -1;
    int idx = -1;

    while(l<= r){
        int m = l + (r - l) /2;
        if(nums[m] == target){
            idx = m;
            if(is_true){
                r = m -1;
            }
            else{
                l = m + 1;
            }
        }
        else if(nums[m] < target){
            l = m + 1;
        }
        else{
            r = m-1;
        }
    }

    return idx;
}
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        return {binarySearch(nums ,target,true) , binarySearch(nums,target,false)};
    }
};